selfdrive/manager/manager.py

```diff
@@ -0,0 +1,220 @@
+    spinner.update_progress(perc, 100.)
+    p.prepare()
+
+
+def cleanup_all_processes(signal, frame):
+  cloudlog.info("caught ctrl-c %s %s" % (signal, frame))
+
+  if EON:
+    pm_apply_packages('disable')
+
+  for p in managed_processes.values():
+    p.stop()
+
+  cloudlog.info("everything is dead")
+
+
+def manager_thread():
+  cloudlog.info("manager start")
+  cloudlog.info({"environ": os.environ})
+
+  # save boot log
+  subprocess.call("./bootlog", cwd=os.path.join(BASEDIR, "selfdrive/loggerd"))
+
+  ignore = []
+  if os.getenv("NOBOARD") is not None:
+    ignore.append("pandad")
+  if os.getenv("BLOCK") is not None:
+    ignore += os.getenv("BLOCK").split(",")
+
+  # start offroad
+  if EON and "QT" not in os.environ:
+    pm_apply_packages('enable')
+    start_offroad()
+
+  started_prev = False
+  params = Params()
+  sm = messaging.SubMaster(['deviceState'])
+  pm = messaging.PubMaster(['managerState'])
+
+  while True:
+    sm.update()
+    not_run = ignore[:]
+
+    if sm['deviceState'].freeSpacePercent < 5:
+      not_run.append("loggerd")
+
+    started = sm['deviceState'].started
+    driverview = params.get("IsDriverViewEnabled") == b"1"
+    ensure_running(managed_processes.values(), started, driverview, not_run)
+
+    # trigger an update after going offroad
+    if started_prev and not started:
+      os.sync()
+      managed_processes['updated'].signal(signal.SIGHUP)
+
+    started_prev = started
+
+    running_list = ["%s%s\u001b[0m" % ("\u001b[32m" if p.proc.is_alive() else "\u001b[31m", p.name)
+                    for p in managed_processes.values() if p.proc]
+    cloudlog.debug(' '.join(running_list))
+
+    # send managerState
+    msg = messaging.new_message('managerState')
+    msg.managerState.processes = [p.get_process_state_msg() for p in managed_processes.values()]
+    pm.send('managerState', msg)
+
+    # Exit main loop when uninstall is needed
+    if params.get("DoUninstall", encoding='utf8') == "1":
+      break
+
+
+def main(spinner):
+  params = Params()
+  params.manager_start()
+
+  default_params = [
+    ("CommunityFeaturesToggle", "0"),
+    ("CompletedTrainingVersion", "0"),
+    ("IsRHD", "0"),
+    ("IsMetric", "0"),
+    ("RecordFront", "0"),
+    ("HasAcceptedTerms", "0"),
+    ("HasCompletedSetup", "0"),
+    ("IsUploadRawEnabled", "1"),
+    ("IsLdwEnabled", "1"),
+    ("LastUpdateTime", datetime.datetime.utcnow().isoformat().encode('utf8')),
+    ("OpenpilotEnabledToggle", "1"),
+    ("VisionRadarToggle", "0"),
+    ("LaneChangeEnabled", "1"),
+    ("IsDriverViewEnabled", "0"),
+  ]
+
+  # set unset params
+  for k, v in default_params:
+    if params.get(k) is None:
+      params.put(k, v)
+
+  # is this dashcam?
+  if os.getenv("PASSIVE") is not None:
+    params.put("Passive", str(int(os.getenv("PASSIVE"))))
+
+  if params.get("Passive") is None:
+    raise Exception("Passive must be set to continue")
+
+  if EON:
+    update_apks()
+
+  manager_init(spinner)
+  manager_prepare(spinner)
```

Comment:
merge these? unclear how these are different

Reply:
the top half of this main function is like a third init too

Reply:
Init and prepare are different. prepare calls prepare on all the proceses, but doesn't do anything to the state in manager.